### PR TITLE
Json prop tracker

### PR DESCRIPTION
**Pull request by donomii**
### What's being changed:

Replace old prop tracker with one using json format.  This fixes a bug, and makes the file future proof.

I'm considering whether I should use "save temp and move" to update the properties file, so it doesn't get corrupted during a sudden power-off

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+	} else {
+
+		t.data.BucketedData[propName] = make(map[int]int, 64)
+		t.data.BucketedData[propName][int(bucketId)] = 1
+	}
+
+	return nil
+}
+
+func (t *JsonPropertyLengthTracker) bucketFromValue(value float32) int {
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:108]: could you explain this?

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+	BucketedData map[string]map[int]int
+}
+
+type JsonPropertyLengthTracker struct {
+	path string
+	data PropLenData
+	sync.Mutex
+}
+
+func NewJsonPropertyLengthTracker(path string) (*JsonPropertyLengthTracker, error) {
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:33]: could you add a docstring on what this class does?

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+	bucket, ok := t.data.BucketedData[propName]
+	if !ok {
+		return 0, nil // Needed for backwards compatibility
+	}
+
+	sum := float32(0)
+	totalCount := float32(0)
+
+	for i := -1; i <= 64; i++ {
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:140]: the 64 seems to be a magic number, could you create a constant?

Comment (dirkkul) [reply adapters/repos/db/inverted/new_prop_length_tracker.go:140]: why do we start at -1?

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	err = os.WriteFile(t.path, bytes, 0o666)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *JsonPropertyLengthTracker) FlushBackup() error {
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:201]: have one function with a "backup"-bool parameter?

File: adapters/repos/db/inverted/prop_length_tracker_test.go [hunk 5]
```diff
@@ -205,6 +205,56 @@ func Test_PropertyLengthTracker_Persistence(t *testing.T) {
...
+		// later page
+		actualMeanForProp20 := float32(1+4+3+17+25) / 5.0
+		res, err = secondTracker.PropertyMean("prop_19")
+		require.Nil(t, err)
+		assert.InEpsilon(t, actualMeanForProp20, res, 0.1)
+	})
+}
+
+// Testing the switch from the old property length tracker to the new one
+
```
Comment (dirkkul) [inline adapters/repos/db/inverted/prop_length_tracker_test.go:251]: remove emtpy line?

File: adapters/repos/db/inverted/prop_length_tracker_test.go [hunk 6]
```diff
@@ -205,6 +205,56 @@ func Test_PropertyLengthTracker_Persistence(t *testing.T) {
...
+func TestFormatConversion(t *testing.T) {
```
Comment (dirkkul) [inline adapters/repos/db/inverted/prop_length_tracker_test.go:252]: are there any more edge cases that we should test?

File: adapters/repos/db/vector/hnsw/vector_cache_prefiller.go [hunk 7]
```diff
@@ -123,13 +123,13 @@ func (pf *vectorCachePrefiller[T]) logLevel(level, count int, before time.Time)
 }
 
 func (pf *vectorCachePrefiller[T]) logTotal(count, limit int, before time.Time) {
-	pf.logger.WithFields(logrus.Fields{
+	/*pf.logger.WithFields(logrus.Fields{
```
Comment (dirkkul) [inline adapters/repos/db/vector/hnsw/vector_cache_prefiller.go:126]: remove if we don't need it?

File: adapters/repos/db/inverted/prop_length_tracker_test.go [hunk 8]
```diff
@@ -73,7 +73,7 @@ func Test_PropertyLengthTracker(t *testing.T) {
```
Comment (dirkkul) [inline adapters/repos/db/inverted/prop_length_tracker_test.go:73]: can you add the  "Test_PropertyLengthTracker_Overflow"-test for the new tracker?

File: adapters/repos/db/vector/hnsw/vector_cache_prefiller.go (hunk 7 above)
Comment (etiennedi) [reply adapters/repos/db/vector/hnsw/vector_cache_prefiller.go:126]: Please don't remove this. This is very helpful for CS when doing manual upgrades, etc. to see when the vector cache is ready.

Comment (donomii) [reply adapters/repos/db/vector/hnsw/vector_cache_prefiller.go:126]: It was spamming my logs when I was debugging, I'll put it back in

File: adapters/repos/db/inverted/new_prop_length_tracker.go (hunk 3 above)
Comment (donomii) [reply adapters/repos/db/inverted/new_prop_length_tracker.go:140]: The answer to these requires a small write-up, or blog post, or something.  Do we have a place to put these?  I feel like we should be writing some extra commentary alongside our work

File: adapters/repos/db/inverted/prop_length_tracker_test.go (hunk 8 above)
Comment (donomii) [reply adapters/repos/db/inverted/prop_length_tracker_test.go:73]: The new tracker can't overflow, so there's no need to test it.

It's one of the good features of the new tracker.

**Comment by donomii** [comment]
Timings show this implementation is a bit slower, I'll look into it:



FIQA

Old format

Import 

________________________________________________________
Executed in   17.47 secs    fish           external
   usr time    2.43 secs    0.27 millis    2.43 secs
   sys time    0.37 secs    2.10 millis    0.36 secs


Objects imported: 57638

Query Latencies 
-------------------- 
Min:      1.7835ms
Mean:   7.442529ms
Max:   23.420584ms
p50:    7.035708ms
p90:   11.268375ms
p99:   15.783542ms

nDCG score: 0.2136, hits at 1: 878, hits at 5: 2006

________________________________________________________
Executed in   45.41 secs    fish           external
   usr time    1.89 secs    0.23 millis    1.89 secs
   sys time    0.74 secs    1.91 millis    0.74 secs


New format, 64 buckets

________________________________________________________
Executed in   13.47 secs    fish           external
   usr time    2.46 secs    0.28 millis    2.45 secs
   sys time    0.37 secs    2.41 millis    0.37 secs



Objects imported: 57638

Query Latencies 
-------------------- 
Min:    1.649917ms
Mean:  10.638113ms
Max:   65.440917ms
p50:    8.554958ms
p90:   18.355917ms
p99:   40.349583ms

nDCG score: 0.2721, hits at 1: 1201, hits at 5: 2691

________________________________________________________
Executed in   64.40 secs    fish           external
   usr time    1.86 secs    0.27 millis    1.86 secs
   sys time    0.74 secs    2.26 millis    0.74 secs


Unlimited buckets

________________________________________________________
Executed in   13.47 secs    fish           external
   usr time    2.45 secs    0.35 millis    2.45 secs
   sys time    0.35 secs    2.58 millis    0.35 secs

Query Latencies 
-------------------- 
Min:    1.661458ms
Mean:  11.574849ms
Max:   66.146125ms
p50:     9.33775ms
p90:   20.397208ms
p99:   42.703542ms

nDCG score: 0.2540, hits at 1: 1076, hits at 5: 2419

________________________________________________________
Executed in   70.08 secs    fish           external
   usr time    1.90 secs    0.27 millis    1.90 secs
   sys time    0.74 secs    2.31 millis    0.74 secs

File: adapters/repos/db/inverted/new_prop_length_tracker.go (hunk 1 above)
Comment (donomii) [reply adapters/repos/db/inverted/new_prop_length_tracker.go:171]: added comment

File: adapters/repos/db/inverted/new_prop_length_tracker.go (hunk 3 above)
Comment (donomii) [reply adapters/repos/db/inverted/new_prop_length_tracker.go]: added comment

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+//
+// These buckets are then recorded to disk.  The original implementation was a binary format where all the data was tracked using manual pointer arithmetic.  The new version tracks the statistics in a go map, and marshals that into JSON before writing it to disk.  There is no measurable difference in speed between these two implementations while importing data, however it appears to slow the queries by about 15% (while improving recall by ~25%).
+//
+// The new tracker is exactly compatible with the old format to enable migration, which is why there is a -1 bucket.  Altering the number of buckets or their values will break compatibility.
+//
+// Set UnlimitedBuckets to true for precise length tracking
+
+// NewJsonPropertyLengthTracker creates a new tracker and loads the data from the given path.  If the file is in the old format, it will be converted to the new format.
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:58]: ver nice explanation, thanks 👍

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+			}
+
+			propertyNames := plt.PropertyNames()
+			data = PropLenData{make(map[string]map[int]int), make(map[string]int), make(map[string]int)}
+			// Loop over every page and bucket in the old tracker and add it to the new tracker
+			for _, name := range propertyNames {
+				data.BucketedData[name] = make(map[int]int, 64)
+				data.CountData[name] = 0
+				data.SumData[name] = 0
+				for i := 0; i <= 64; i++ {
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:93]: introduce a constant for the 64?

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+	return float32(sum) / float32(count), nil
+}
+
+// returns totalPropertyLength, totalCount, average propertyLength = sum / totalCount, total propertylength, totalCount, error
+func (t *JsonPropertyLengthTracker) PropertyTally(propName string) (int, int, float64, error) {
+	t.Lock()
+	defer t.Unlock()
+	sum, ok := t.data.SumData[propName]
+	if !ok {
+		return 0, 0, 0, nil
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:226]: return an error here?

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+	}
+	count, ok := t.data.CountData[propName]
+	if !ok {
+		return 0, 0, 0, nil
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:230]: also, should we return an error here?

If that doesn't make sense, can you remove the error from the return value? Then there is no codepath that returns an error

File: adapters/repos/db/shard_write_delete.go [hunk 13]
```diff
@@ -53,6 +54,27 @@ func (s *Shard) deleteObject(ctx context.Context, id strfmt.UUID) error {
...
+	if err != nil {
+		return errors.Wrap(err, "get existing object from binary")
+	}
+
+	properties := obj.Object.Properties
+
+	// Remove each property from the property length tracker.
+	for _, prop := range properties.(map[string]interface{}) {
+		switch prop.(type) {
+		case string:
```
Comment (dirkkul) [inline adapters/repos/db/shard_write_delete.go:67]: isn't it `text` now?

File: test/benchmark_bm25/cmd/query.go [hunk 14]
```diff
@@ -89,9 +89,12 @@ var queryCmd = &cobra.Command{
 			}
 
 			result, err := bm25Query.Do(context.Background())
+
+			result, err = bm25Query.Do(context.Background())
```
Comment (dirkkul) [inline test/benchmark_bm25/cmd/query.go:93]: why double this?

File: test/benchmark_bm25/gather_stats.sh [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+#./FlameGraph/stackcollapse-go.pl "${NAME}_cpu.txt" | ./FlameGraph/flamegraph.pl > "${NAME}_flame.svg" &
+open "${NAME}_heap.png"
+open "${NAME}_profile.png"
+open "${NAME}_allocs.png"
+
+
+open "${NAME}_flame.svg"
+
+
```
Comment (dirkkul) [inline test/benchmark_bm25/gather_stats.sh:33]: cleanup or delete? :)

File: adapters/repos/db/inverted/new_prop_length_tracker.go [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+		if os.IsNotExist(err) {
+			t.Flush(false)
+			return t, nil
+		}
+		return nil, err
+	}
+	t.path = path
+
+	var data PropLenData
+	if err := json.Unmarshal(bytes, &data); err != nil {
```
Comment (dirkkul) [inline adapters/repos/db/inverted/new_prop_length_tracker.go:78]: what happens when weavaite crashes/gets shut down during the conversion?

**Review by dirkkul — APPROVED** [review]
Can you also run the chaos pipeline for this PR?

File: adapters/repos/db/inverted/new_prop_length_tracker.go (hunk 12 above)
Comment (donomii) [reply adapters/repos/db/inverted/new_prop_length_tracker.go:230]: Backwards compatibility :(

File: test/benchmark_bm25/cmd/query.go (hunk 14 above)
Comment (donomii) [reply test/benchmark_bm25/cmd/query.go:93]: Leftover from debugging

File: adapters/repos/db/inverted/new_prop_length_tracker.go (hunk 11 above)
Comment (donomii) [reply adapters/repos/db/inverted/new_prop_length_tracker.go:226]: Backwards compatibility :(

File: adapters/repos/db/inverted/new_prop_length_tracker.go (hunk 10 above)
Comment (dirkkul) [reply adapters/repos/db/inverted/new_prop_length_tracker.go:93]: Why don't we call it MAX_BUCKETS and add a the reason as a comment

File: adapters/repos/db/inverted/new_prop_length_tracker.go (hunk 11 above)
Comment (dirkkul) [reply adapters/repos/db/inverted/new_prop_length_tracker.go:226]: ah bummer, a short comment would be helpful for the next one.

What does it mean if a prop is not found in the SumData? That no object was added that has that property defined?

Comment (donomii) [reply adapters/repos/db/inverted/new_prop_length_tracker.go]: Yes, the props only get created if there is data for them

File: adapters/repos/db/shard_write_delete.go (hunk 13 above)
Comment (donomii) [reply adapters/repos/db/shard_write_delete.go]: done

Comment (dirkkul) [reply adapters/repos/db/shard_write_delete.go]: afaik string does not exist anymore, pinging @aliszka to confirm?
